snowdrop-block/src/Snowdrop/Block/State.hs

```diff
@@ -51,63 +59,85 @@ instance Buildable blockRef => Buildable (BlockRef blockRef) where
+inmemoryBlkStateConfiguration cfg validator mkProof mkBlock = fix $ \this ->
     BlkStateConfiguration {
       bscConfig = cfg
     , bscExpand = \rawBlock -> do
-          blkPayload <- liftERoComp $ expandUnionRawTxs mkProof (gett rawBlock)
+          blkPayload <- liftERoComp $ upcastEffERoComp $ expandUnionRawTxs mkProof (gett rawBlock)
           pure (mkBlock rawBlock blkPayload)
-    , bscApplyPayload = \(gett @_ @[SomeTx id value (RContains txtypes)] -> txs) -> do
-          undos <- forM txs $ \tx -> do
-                      liftERoComp $ applySomeTx (runValidator validator) tx
-                      modifyRwCompChgAccum (CAMChange $ applySomeTx txBody tx)
+    , bscApplyPayload = \(gett @_ @[SomeTx (BlkProcConstr txtypes xs)] -> txs) -> do
+          undos <- forM txs $ \smtx -> usingSomeTx smtx $ \tx -> do
+                      liftERoComp $ upcastEffERoComp $ runValidator validator tx
+                      applySomeTx (modifyRwCompChgAccum . CAMChange . hupcast . txBody) smtx
           let mergeUndos (Undo cs1 _) (Undo cs2 sn2) = flip Undo sn2 <$> mappendChangeSet cs1 cs2
           case reverse undos of
-              []     -> pure $ Undo def BS.empty
+              []     -> pure $ Undo def def
               f:rest -> either throwLocalError pure $ foldM mergeUndos f rest
     , bscApplyUndo = void . modifyRwCompChgAccum . CAMRevert
     , bscStoreBlund = \blund -> do
           let blockRef = unCurrentBlockRef $ bcBlockRef cfg (blkHeader $ buBlock blund)
-          let chg = ChangeSet $ M.singleton (inj $ blockRef) (New $ inj blund)
-          void $ modifyRwCompChgAccum $ CAMChange chg
-    , bscRemoveBlund = \blockRef ->
-        void $ modifyRwCompChgAccum $ CAMRevert $ Undo (ChangeSet $ M.singleton (inj blockRef) Rem) BS.empty
-    , bscGetBlund = liftERoComp . queryOne
-    , bscBlockExists = liftERoComp . queryOneExists
-    , bscGetTip = liftERoComp (queryOne TipKey)
-                    >>= maybe (throwLocalError @(BlockStateException id) TipNotFound) (pure . unTipValue)
+          let chg = hChangeSetFromMap @(BlundMap blkType) $ M.singleton blockRef (New blund)
+          void $ modifyRwCompChgAccum $ CAMChange $ hupcast chg
+    , bscRemoveBlund = \blockRef -> do
+          let chg = hChangeSetFromMap @(BlundMap blkType) $ M.singleton blockRef Rem
+          void $ modifyRwCompChgAccum $ CAMRevert $ Undo (hupcast chg) def
+    , bscGetBlund = liftERoComp . queryOne @(BlundMap blkType)
+    , bscBlockExists = liftERoComp . queryOneExists @(BlundMap blkType)
+    , bscGetTip = do
+          tipMb <- liftERoComp (queryOne @(TipMap blkType) TipKey)
+          maybe (throwLocalError @BlockStateException TipNotFound) (pure . unTipValue) tipMb
     , bscSetTip = \newTip' -> do
-          let newTip = inj $ TipValue newTip'
-          let tipChg = \cons -> ChangeSet $ M.singleton (inj TipKey) (cons newTip)
-          oldTipMb <- liftERoComp $ queryOne TipKey
+          void $ bscGetTip this -- to check existence
```

Comment:
Is it  by design that with tip absence an exception will be thrown (previously it wasn't so).

Reply:
yep, it's by design